src/probes/bert_probe.c

```diff
@@ -0,0 +1,145 @@
+#include "nica/b64enc.h"
+
+#include "config.h"
+
+static const char bert_record_file[] = "/sys/firmware/acpi/tables/data/BERT";
+static const char telem_record_class[] = "org.clearlinux/debug/bert";
+static uint32_t severity = 2;
+static uint32_t payload_version = 1;
+
+
+void print_usage(char *prog)
+{
+        printf("%s: Usage\n", prog);
+        printf("  -f,  --config_file    Configuration file. This overides the other parameters\n");
+        printf("  -h,  --help           Display this help message\n");
+        printf("  -V,  --version        Print the program version\n");
+}
+
+static int allocate_payload_buffer(char **payload)
+{
+        int ret = 0;
+
+        *payload = (char *) malloc(MAX_PAYLOAD_SIZE);
+
+        if (*payload == NULL) {
+                goto out1;
+        }
+
+        *payload = memset(*payload, 0, MAX_PAYLOAD_SIZE);
+        ret = 1;
+
+out1:
+        return ret;
+}
+
+int main(int argc, char **argv)
+{
+        struct telem_ref *tm_handle = NULL;
+        char *classification = (char *) telem_record_class;
+        char *payload = NULL;
+
+        // Following vars are for arg parsing.
+        int c;
+        char *cfile = NULL;
+        int opt_index = 0;
+        int ret = 0;
+
+        struct option opts[] = {
+                { "cfile", 1, NULL, 'f' },
+                { "help", 0, NULL, 'h' },
+                { "version", 0, NULL, 'V' },
+                { NULL, 0, NULL, 0 }
+        };
+
+        while ((c = getopt_long(argc, argv, "f:hHV", opts, &opt_index)) != -1) {
+                switch (c) {
+                        case 'f':
+                                if (asprintf(&cfile, "%s", (char *) optarg) < 0) {
+                                        exit(EXIT_FAILURE);
+                                }
+
+                                struct stat buf;
+                                ret = stat(cfile, &buf);
+
+                                /* check if the file exists and is a regular file */
+                                if (ret == -1 || !S_ISREG(buf.st_mode)) {
+                                        exit(EXIT_FAILURE);
+                                }
+
+                                tm_set_config_file(cfile);
+                                break;
+                        case 'h':
+                                print_usage(argv[0]);
+                                exit(EXIT_SUCCESS);
+                        case 'V':
+                                printf(PACKAGE_VERSION "\n");
+                                exit(EXIT_SUCCESS);
+                        case '?':
+                                exit(EXIT_FAILURE);
+                }
+        }
+
+        if (!(ret = allocate_payload_buffer(&payload))) {
+                printf("Unable to allocate more memory.\n");
+                goto fail;
+        }
+
+        if(!(ret = nc_b64enc_filename(bert_record_file, payload, MAX_PAYLOAD_SIZE))) {
+                printf("Failed to read payload from: %s\n", bert_record_file);
```

Comment:
How often will this error be hit? Things to consider:

* If the BERT file may not exist at bootup, then perhaps return a special non-zero exit code (e.g. `2`) that bertprobe.service will consider "non-fatal" by specifying (for example) `SuccessExitStatus=2` in the unit file.
* If the BERT file will always exist at bootup (and prior to the bertprobe running), then making the exit status fatal sounds reasonable. If there is a potential race condition where the bertprobe tries the read the BERT file before it is created, that's a separate issue :-)
* Having not looked through the `nc_b64enc_filename` code much, should other errors it raises be handled specially by this probe?

The main reason why I am commenting about this is because the journal probe will report any "fatal" service errors. So this new probe should only return a fatal error if telemetry should be sent for it. The other error conditions handled in the code look appropriately fatal, so I'm only concerned about this one.

Reply:
A BERT file will be available only if on a previous boot there was an error, so most of the time this probe will fail (as coded right now)

I'm going to use a 'ConditionPathExists=/sys/firmware/acpi/tables/data/BERT' line. I feel like the probe should not run at all if the condition is not present.

Reply:
Great! I didn't consider that option. This seems clean to me.